### DISCUSSION TITLE
🎉 New Release: Mobile floating button, navigation bar access, positioning and item visibility by platform (v1.7)

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#62] [Announcements]
The v1.7 release is all about about the mobile experience!

### Installation

Install via Community Plugins.

### Position:

- Position toolbars by platform: Desktop or Mobile
- Select the `Floating button` position (mobile only) to:
	- Add a button to the bottom `left` or `right` in the editor
	- When tapped, it opens your toolbar as a "menu toolbar"
	- Select your own icon (see new setting below)
	- Customize the look via the Style Settings plugin
- A new position = `Hidden` option completely hides the toolbar, which is useful if you just want to access toolbars from the navigation bar (see below).

<img src="https://raw.githubusercontent.com/chrisgurney/obsidian-note-toolbar/master/docs/images/demo_mobile_floating_button.gif?raw=1" width="400px"/>

- Access the toolbar from the navigation bar (mobile only):
	- On mobile, go to Obsidian settings: `Appearance > Ribbon menu > Manage` and select `Note Toolbar` (or select it from the ribbon menu) to add it to your navigation bar
	- Select your own icon (see new setting below)

<img src="https://raw.githubusercontent.com/chrisgurney/obsidian-note-toolbar/master/docs/images/mobile_nav_ribbon_bar.png" width="400px"/>

- Quickly toggle between the `Top (fixed)` and `Below Properties` positions after right-clicking on a toolbar

![CleanShot 2024-05-10 at 10 02 23@2x](https://github.com/chrisgurney/obsidian-note-toolbar/assets/17337703/6d943799-79ce-4e55-9341-8a090fbedc0b)

### New mobile-only settings (below the folder mappings section) to:
- Change the icon in the navigation bar / floating button
- Toggle an `Edit toolbar` link in the menu toolbar, for convenience

### Visibility:
- When adding toolbar items, set whether the icons, and/or label is shown (or neither) on desktop or mobile
    - Menu toolbars show the icon and label, or use the tooltip if a label is not set.

### Commands:
- New `Open toolbar settings` command that opens the settings for the toolbar on the present note

### Style Settings plugin:
- Reorganized styles under sub-headings to hopefully make easier to find: Colors, Positioning, Sizing and Spacing, and Style
- Added settings to adjust toolbar and item spacing

![CleanShot 2024-05-07 at 15 30 16@2x](https://github.com/chrisgurney/obsidian-note-toolbar/assets/17337703/d0a4a46c-f4a1-4dee-a744-b5e8c0d988b8)

## Feedback Wanted 🤝

I'm looking for feedback, as well as help with testing on Android, iOS, and tablets:

1. What do you think of the look and feel of the floating button, icon, and menu bar?
2. Would you use the floating button on a tablet?
3. How well does the button work for you? Are you experiencing any issues with it? 
    - I've experienced an issue on iOS where the links in the menu bar no longer work after using the menu for a time, but still need steps to reproduce to understand what's going on there.
4. Where is the button showing for Android devices?

**Comment by FelipeRearden** [comment]
Hello @chrisgurney !!!!

> I'm looking for feedback, as well as help with testing on Android, iOS, and tablets:

I am in 👍 
I am gonna focus on tests of the beta version on iPadOS WITH and WITHOUT an external mouse and keyboard.

**Reply by chrisgurney** [reply]
Thank you! Much appreciated!

**Reply by FelipeRearden** [reply]
> Would you use the floating button on a tablet?

Chris, just a question okay… `Floating Button: right or left` should appear on tablets on `1.7.7` ?

**Reply by chrisgurney** [reply]
I believe they should, as it's checking Obsidian's isMobile variable that I think includes tablets.

What are you experiencing?

**Reply by FelipeRearden** [reply]
> What are you experiencing?

On Vanilla Vault with fresh install of version`1.7.7`  I cannot see `Floating Button: right or left`. When `Floating` is selected, it is the same behavior as `hidden` 

ℹ️ `hidden` is working perfect on Reading Mode and Live preview;
ℹ️ `Top: fixed` and `Below Properties` are working perfect on Reading Mode and Live preview;

**Reply by chrisgurney** [reply]
Strange. On my iPad, I see the floating button in portrait and landscape mode.

~~However, when tapping the button, the menu opens in the upper left, which is obviously incorrect.~~ Fixed, but more than one menu opens (on top of the last) when tapping the button again.

Can you share more about your configuration @FelipeRearden ?

**Reply by FelipeRearden** [reply]
> Strange. On my iPad, I see the floating button in portrait and landscape mode.

Working perfect on beta version `1.7.10`. It was a mistake on my end, forgive for the mistake :)

> Fixed, but more than one menu opens (on top of the last) when tapping the button again.

Same here. Maybe you could use the info that you gave me here #23 to evaluate if the menu is open and `do nothing` if the menu is open. Just an idea that might be a dumb one :)

**Reply by chrisgurney** [reply]
Thanks for confirming!

I'm curious:

1. Would your preference be to keep toolbars as-is on tablet, and maybe use the button on the phone?

2. What's your opinion on where the button is located, and its size on the iPad?

**Reply by chrisgurney** [reply]
@FelipeRearden Just saw your other comment, but would still love your opinion on the above.

**Reply by FelipeRearden** [reply]
Just to remember that I use my Obsidian Mobile on 12.9 iPad as a Desktop: landscape + keyboard and mouse:
- I even hide the mobile toolbar to same screen space;

I will start to use Obsidian Mobile on iPadMini as a big iPhone on portait mode. That’s where I am thinking about he `floating button` but it is a WIP for me.  
 
I dont use Obsidian Mobile on iPhone. I dont have an iPhone right now :(

> 1. Would your preference be to keep toolbars as-is on tablet, and maybe use the button on the phone?

My toolbar workflow continuous to be `Top: fixed` on iPad;

But looks the best choice to use `floating button` on iPhone and iPadMini;

> 2. What's your opinion on where the button is located, and its size on the iPad?

I am on 12.9 iPad and it is okay. But I think we need a better feedback on smaller iPad sizes.

**Reply by chrisgurney** [reply]
OK thanks for letting me know! 

Please don't feel like you have to go out of your way to test this @FelipeRearden! I'm sure we'll find a bunch of other volunteers out there. (I'll announce the beta shortly in the Discord to get more eyeballs on this, to start.)

**Reply by FelipeRearden** [reply]
Thanks Chris, I really appreciate your attention !!! 

As always I am here for testing the new features ! 

Let’s see how the other volunteers feel about the new features :)

**Comment by FelipeRearden** [comment]
> I'm looking for feedback, as well as help with testing on Android, iOS, and tablets:
> 
> 1. What do you think of the look and feel of the floating button, icon, and menu bar?

I am not the standard Mobile user but it looks an amazing addition to interact with Obsidian Mobile on phones and tablets (7/8 inches).

> Would you use the floating button on a tablet?

Yes. Specially for iPad Mini users on Portait Mode. The subset of workflows here is almost endless. 

Let’s see how the community behaves. But from my standpoint it is a big winner for both mobile.

**Reply by chrisgurney** [reply]
@FelipeRearden I'm not sure if you've tried the `Focus` command with the floating button, but I've just added a fix to the latest beta where `Focus` will trigger the floating button's menu (it wasn't before). You can then use the arrow keys within the menu (like in any other context menu in the app).

**Reply by FelipeRearden** [reply]
No I didnt @chrisgurney ! I am happy to know that `Focus` is working on a floating button now :)

**Comment by chrisgurney** [comment]
Quick update on this release: I have a user who's identified an issue with the floating button menu on their device (iPhone 15 Pro). Since I don't have said device, I'm in the process of trying to narrow down if this is a configuration/migration problem, or something else entirely.

**Comment by dlangendorf** [comment]
Hey, Chris. This is an amazing plugin. Thanks for all the hard work. I was wondering: I have the toolbar set up on a laptop where my vault is stored. I also want to use it on an iPhone (and probably an iPad), but I can't get it to work (probably user error). I couldn't tell from the documentation. Do I need to set it up on every device I want to use it on, or will one setup work across devices? I've made sure set up is exactly the same, but I must be doing something wrong. Any thoughts as to using toolbar on multiple devices? Today's daily note, set up on laptop, doesn't contain notebook on mobile and I can't find where I'm messing up. Anyway, just wondering. And again thanks for all your work.

**Reply by dlangendorf** [reply]
Chris, I figured it out. I uninstalled, closed out of Obsidian, restarted, redownloaded Toolbar and this time I was able to map folders. I look forward to seeing where you might go with this ... and what I can do with it.

**Reply by chrisgurney** [reply]
@dlangendorf Oh, good to hear! I wasn't sure what resource to send your way, nor did I know how you were syncing your devices. Glad to hear you sorted it out then!

I would be curious to know what your experience with using the plugin's like, especially with respect to setting it up for the first time: What made sense, what didn't, and what you might like to see in the future.

**Comment by guycalledxan** [comment]
[First post so I'll say: Super cool plugin. I'd had a hack in some of my note templates to make an embedded toolbar. I was super pleased to switch when you released this]

Since this floating mobile button came out, I've switched to that format on mobile (iOS) and prefer it. However, one of the first things I tried to look for was a way to get it to position top-right. I can live with it bottom-right, but even after using it for a while, I would prefer it top-right. I think partly because when it's bottom-right, it jumps around as the keyboard shows/hides. Top-right would keep it in a more predicable place.

Would it be possible to add this (and I guess top-left, for left-handed people - so all corners are covered)? It seemed like an odd omission, so I half-suspect there's already a known reason that didn't work, but I figure I might as well ask, in case it's doable.

**Reply by chrisgurney** [reply]
@guycalledxan Thanks for the kind comment!

Via the [Style Settings plugin](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Style-Settings-plugin-support) you can adjust the position of the floating button, under the setting `Button Y position`. Setting the value to a lower number should put it higher on the screen. (Note that you may have to use a different setting per device, as it's relative to screen size.) 

Perhaps give that a go and see if you can get it in a spot that works for you. (I'd love to see a screenshot if you do this.)

Floating buttons are typically on the bottom of the screen (closer to the thumb for mobile access). I can take a look at positioning it at the top but my suspicion is that I might not be able to find a good spot for it: there's the varying length of note titles to contend with (there could be overlap); or if I tried a lower spot, there's a varying number of Properties per note. (Button placement is relative to the editor window.)

Hope this helps!

**Reply by guycalledxan** [reply]
Cheers for the reply. :)

I'd fiddled with the style settings a bit, but I guess before I started to think about the button position. I'd forgotten about them.

I just tried the Y position and it kinda works (iPhone 15 Pro, whatever the latest update version is). Enough that I'll keep it like this for now.

It's better than before but it still leaves two problems (one old, one new). The button still jumps around as I open the keyboard, though relatively less so with a smaller % value. The jumping feeds into the new problem: I also have the Tab Title Bar enabled (for use with Commander). That seems to draw over the window area the % derives from, and draws over the button. It's also as tall as the button. I've found that any % under 30 will cause the button to be pushed under the tab toolbar when the keyboard opens. Less than 25% and it disappears under the tab toolbar. At 30% it's in a slightly weird place onscreen, but will be unobscured enough to click when the keyboard is open.

I've been contemplating moving everything from the tab toolbar to Note Toolbar but haven't done so yet, and haven't properly decided. They're my "default, have everywhere" actions. Whereas Note toolbar are my note-type specific actions. I think unless Note Toolbar allowed for child toolbars to inherit from and modify a parent (at least a global one), I'd prefer my current setup.

> I can take a look at positioning it at the top but my suspicion is that I might not be able to due to find a good spot for it: there's the varying length of note titles to contend with

I wonder if you could add an option to have the relative y % offset be from the top of whatever area it currently is? The opposite edge (or corner for x,y), in other words. Then I'd probably set the y offset to 10 or 90 (depending on how it's calculated), to get it just below the Tab Toolbar. Presumably, then, that edge wouldn't change position as the keyboard pops up, in the same way as the edge nearest the keyboard does? I'm not really sure how feasible what I'm suggesting is though - I've only done very basic dynamic ui stuff before. I'd be the client/manager who thinks they have a good idea but is actually just full of s*. :P

Btw, exposing stuff with style settings was a _really_ smart feature. I've not seen other plugins do that yet. I wish other things, like Reminders did. It allowed me to tweak my desktop toolbar exactly how I wanted, whereas I was hacking the css in the very first release.

![IMG_1557](https://github.com/chrisgurney/obsidian-note-toolbar/assets/42411202/aaa02da4-487e-449f-989b-180c73e1a199)
![IMG_1558](https://github.com/chrisgurney/obsidian-note-toolbar/assets/42411202/0223142b-edf4-4900-8b0f-dc212ed819a7)

**Reply by chrisgurney** [reply]
> I wonder if you could add an option to have the relative y % offset be from the top of whatever area it currently is? 

It is actually offset from the top, hence why higher numbers are at the bottom. But you raise a good point: I'm not sure off the top of my head why the button's jumping around when the keyboard appears. Hmm.

As an aside: There's also an amount of `110px` subtracted from the offset, but I'm not 100% sure why. I think I may have cribbed this from the [Things theme](https://github.com/colineckert/obsidian-things), and it seemed to work well.

I'm open to working with a front-end dev on this who has more experience with CSS than me, and trying other options!

**Reply by chrisgurney** [reply]
@guycalledxan FYI the latest beta of Note Toolbar now allows you to float a toolbar at the bottom of notes. I would be interested in knowing if that helps you at all.

**Comment by chrisgurney** [comment]
@guycalledxan 

> I think unless Note Toolbar allowed for child toolbars to inherit from and modify a parent (at least a global one), I'd prefer my current setup.

This is next on my list: The idea is that you'll be able to set a "Toolbar" type for a particular item, and have it be replaced with the items in the selected toolbar. Just to confirm, is this what you meant?

**Reply by guycalledxan** [reply]
Sorry, it might be "Monday morning, not enough coffee" but I don't quite follow the description. If I'm guessing at stuff:

An item in the context of Note Toolbar is what's defined using Edit Toolbar>Items?

If that's the case, I think at the moment the only way they can be defined is within a Toolbar (created via the [+ New toolbar] button.

What I infer from your description is that:
1) Items (what I might think of as a toolbar button) will be defined elsewhere ("top level" or "globally" within Note Toolbar's settings) and, as part of the Item definition, assigned to a Toolbar by toolbar name (its identifier). Then,
2) Within the toolbar, there'd also be some kind of further definition of Items (the existing way) where they would add to the toolbar, in addition to the "globally" defined and assigned Items, and (possibly?) override global items with local ones where the names match?

A lot of wild assumptions there on my part but if I'm kinda right, yeah, that'd solve the problem (user story, I guess) I mentioned.

It'd be an elegant way to do it too. Defining one item and assigning it to many toolbars would be a lot more convenient than crating what's basically the same item multiple times in each toolbar. I guess if it's 1 Item to many Toolbars then you don't need (2), the current way at all. In other words, no need for "inheritance".

I guess mostly, I got confused by the line:
> and have it be replaced with the items in the selected toolbar.

**Reply by chrisgurney** [reply]
@guycalledxan Sorry about that, I think I was in a rush when I wrote that. 😅

Here's two new types I'd like to add in v1.9:

### Item Type: _Toolbar_ 

Here's an example:

- Toolbar 1 has these items: A, B, C
- Toolbar 2 has these items: D, E

Let's say I add another _Toolbar_ type item to Toolbar 2 that's just a reference to Toolbar 1:

- Now Toolbar 2 has these items: D, E, A, B, C

### Item Type: _Menu_

Let's say I add a _Menu_ type item to Toolbar 2 instead, that's a reference to Toolbar 1:

- Toolbar 2 will have these items: D, E, F
- When F is clicked, opens a drop-down menu containing: A, B, C

**Reply by guycalledxan** [reply]
Ah, right. That's clear. That seems like a good implementation, and it'd do what I need.

Had I not written the above, I might not have really thought about it. But it might be useful to be able to define a 3rd type: "action" (or "button") and use that directly inside of a Toolbar.

If I wanted to have these at various points in obsidian:

- Toolbar 1 has these items: A, B, C
- Toolbar 2 has these items: A, F

With what you've written, it sounds like it'd be possible to make a Toolbar or Menu with one action inside, which is fine. So I'd probably make "Toolbar 3: A" so I can have a central definition of A. It depends on the UI and how often I'd want to do that of course, but it might be convenient to not have to wrap up the actions in a toolbar.

**Reply by FelipeRearden** [reply]
@chrisgurney just to make sure that I understood correctly…

> Item Type: Toolbar
> 
> Here's an example:
> 
> Toolbar 1 has these items: A, B, C
> Toolbar 2 has these items: D, E
> Let's say I add another Toolbar type item to Toolbar 2 that's just a reference to Toolbar 1:
> 
> Now Toolbar 2 has these items: D, E, A, B, C

Question: Toolbar 2 will SHOW `D, E, A, B, C` after a USER CLICK or as the default view ?

I am asking this based on the example from my FR #38 where (following your example) the the Toolbar 1 will be showed ONLY when the user click the item that is connected labeled as TOOLBAR.

I am asking this because I am example that TB1 and TB2 has 10 items each, I am afraid about a scenario where the Toolbar would have 20 items (for example). 

My question in a few words: `TB1` will replace `TB2` or `TB1` will be added to `TB2` ?

**Reply by chrisgurney** [reply]
@FelipeRearden The idea is that the Toolbar option will _replace_ the items by default (ie., without a click). If you have a lot of items, that's what the `Menu` option can be useful for.

I felt like that was a good compromise to start, and we can perhaps explore other options from there.

**Reply by FelipeRearden** [reply]
100% understood @chrisgurney ! 

> I felt like that was a good compromise to start, and we can perhaps explore other options from there.

🙏 

I am excited to see the community feedback about this item. Because in my humble opinion, for complex toolbars  like the ones that I am thinking about, an option like the one I described in #38 might be a good option. 

As always, thank you very much for this amazing plugin and for let us suggest new ideas for the plugin :)
